File: KD Enabled/filesystem/crypto.py
import os
import hmac
import hashlib
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from typing import Tuple

# ---------------- Constants ----------------
KEY_LEN = 32      # AES-256
HMAC_LEN = 32     # HMAC-SHA256 (digest size)
NONCE_LEN = 8     # AES-CTR nonce length (must be <= 15, 8 is standard)
PBKDF2_ITER = 100_000


# ---------------- Random Bytes ----------------
def random_bytes(n: int) -> bytes:
    """Generate secure random bytes."""
    return get_random_bytes(n)


# ---------------- Key Derivation ----------------
def derive_key(password: str, salt: bytes) -> bytes:
    """Derive a key using PBKDF2 with SHA-256."""
    return hashlib.pbkdf2_hmac(
        "sha256", password.encode("utf-8"), salt, PBKDF2_ITER, KEY_LEN
    )


# ---------------- AES-CTR ----------------
def aes_ctr_encrypt(key: bytes, plaintext: bytes) -> Tuple[bytes, bytes]:
    """Encrypt with AES-CTR. Returns (ciphertext, nonce)."""
    if len(key) != KEY_LEN:
        raise ValueError("Invalid AES key length")

    nonce = random_bytes(NONCE_LEN)
    cipher = AES.new(key, AES.MODE_CTR, nonce=nonce)
    ciphertext = cipher.encrypt(plaintext)
    return ciphertext, nonce


def aes_ctr_decrypt(key: bytes, nonce: bytes, ciphertext: bytes) -> bytes:
    """Decrypt AES-CTR ciphertext."""
    if len(key) != KEY_LEN:
        raise ValueError("Invalid AES key length")
    if len(nonce) != NONCE_LEN:
        raise ValueError(f"Invalid nonce length: got {len(nonce)}, expected {NONCE_LEN}")

    cipher = AES.new(key, AES.MODE_CTR, nonce=nonce)
    return cipher.decrypt(ciphertext)


# ---------------- HMAC ----------------
def compute_hmac(key: bytes, data: bytes) -> bytes:
    """Compute HMAC-SHA256."""
    return hmac.new(key, data, hashlib.sha256).digest()


def verify_hmac(key: bytes, data: bytes, mac: bytes) -> bool:
    """Verify HMAC-SHA256."""
    expected = compute_hmac(key, data)
    return hmac.compare_digest(mac, expected)

File: Password-Only/filesystem/crypto.go
package filesystem

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	KeyLen     = 32 // AES-256
	HMACLen    = 32 // HMAC-SHA256
	NonceLen   = 16 // AES-CTR nonce
	PBKDF2Iter = 100_000
)

// ---------------- Random Bytes ----------------
func RandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	if _, err := io.ReadFull(rand.Reader, b); err != nil {
		return nil, err
	}
	return b, nil
}

// ---------------- Key Derivation ----------------
func DeriveKey(password string, salt []byte) []byte {
	return pbkdf2.Key([]byte(password), salt, PBKDF2Iter, KeyLen, sha256.New)
}

// ---------------- AES-CTR ----------------
func AESCTREncrypt(key, plaintext []byte) (ciphertext, nonce []byte, err error) {
	if len(key) != KeyLen {
		return nil, nil, errors.New("invalid AES key length")
	}
	nonce, err = RandomBytes(NonceLen)
	if err != nil {
		return nil, nil, err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, err
	}

	ct := make([]byte, len(plaintext))
	stream := cipher.NewCTR(block, nonce)
	stream.XORKeyStream(ct, plaintext)
	return ct, nonce, nil
}

func AESCTRDecrypt(key, nonce, ciphertext []byte) ([]byte, error) {
	if len(key) != KeyLen {
		return nil, errors.New("invalid AES key length")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	plain := make([]byte, len(ciphertext))
	stream := cipher.NewCTR(block, nonce)
	stream.XORKeyStream(plain, ciphertext)
	return plain, nil
}

// ---------------- HMAC ----------------
func ComputeHMAC(key, data []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(data)
	return m.Sum(nil)
}

func VerifyHMAC(key, data, mac []byte) bool {
	expected := ComputeHMAC(key, data)
	return hmac.Equal(mac, expected)
}
